src/Protocols/Wishbone/Standard/Hedgehog.hs

```diff
@@ -25,63 +27,184 @@ import Protocols hiding (stallC, circuit)
+--                                    ^ go to next cycle
+nextStateCommonSense _ _ s2m
+  | P.length (filter ($ s2m) [acknowledge, err, retry]) > 1 =
+    Left "More than one termination signal asserted"
+nextStateCommonSense CSVSQuiet WishboneM2S {..} s2m
+  | busCycle && not strobe = Right (True, CSVSInCycleNoStrobe)
+  | busCycle && strobe && writeEnable = Right (False, CSVSWriteCycle)
+  | busCycle && strobe && not writeEnable = Right (False, CSVSReadCycle)
+  | hasTerminateFlag s2m = Left "Termination signals outside of a bus cycle"
+  | otherwise = Right (True, CSVSQuiet)
+nextStateCommonSense CSVSInCycleNoStrobe WishboneM2S {..} _
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | writeEnable = Right (False, CSVSWriteCycle)
+  | not writeEnable = Right (False, CSVSReadCycle)
+  | otherwise = C.error "Should not happen"
+nextStateCommonSense CSVSReadCycle m2s@WishboneM2S {..} s2m@WishboneS2M {..}
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | acknowledge =
+    if responseValid m2s s2m
+      then Right (True, CSVSQuiet)
+      else Left "read-response does not respect SEL"
+  | err || retry = Right (True, CSVSQuiet)
+  | writeEnable = Left "asserted WE while in a read cycle"
+  | otherwise = Right (True, CSVSReadCycle)
+  where
+    responseValid WishboneM2S {busSelect=sel} WishboneS2M {readData=dat} = selectValidData sel dat
+nextStateCommonSense CSVSWriteCycle m2s@WishboneM2S {..} s2m@WishboneS2M {..}
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | not writeEnable = Left "deasserted WE while in a write cycle"
+  | not $ requestValid m2s = Left "write request does not respect SEL"
+  | err || retry = Right (True, CSVSQuiet)
+  | acknowledge =
+    if responseValid s2m
+      then Right (True, CSVSQuiet)
+      else Left "write-response contains defined read-data"
+  | otherwise = Right (True, CSVSWriteCycle)
+  where
+    requestValid WishboneM2S {busSelect=sel, writeData=dat} = selectValidData sel dat
+    responseValid WishboneS2M {readData=dat0} =
+      let dat1 = bitCoerce dat0
+          numBits = natVal @_ @BitVector dat1
+          mask = 2 P.^ numBits - 1
+       in unsafeMask dat1 == mask
+
+selectValidData ::
+  forall a.
+  (KnownNat (BitSize a), BitPack a) =>
+  BitVector (BitSize a `DivRU` 8) ->
+  a ->
+  Bool
+selectValidData byteSelect rawDat =
+  all
+    (== False)
+    [ hasUndefined part
+      | idx <- indices byteSelect,
+        let part = dat ! idx
+    ]
+  where
+    dat :: C.Vec (BitSize a `DivRU` 8) (BitVector 8)
+    dat = case maybeIsX rawDat of
+      Just val -> bitCoerce $ resize (bitCoerce val :: BitVector (BitSize a))
+      Nothing -> C.repeat undefined#
+
+    indices :: forall n. (KnownNat n) => BitVector n -> [Index n]
+    indices bv = [idx | idx <- [0 .. maxBound], bitToBool (bv ! idx)]
+
+validateCommonSense ::
```

Comment:
docs

Reply:
I removed these functions in favour of the validator circuits.